File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int answer = 0;
    
    public int solution(int[] numbers, int target) {
        
        dfs(0,0,0,target, numbers);
        
        return answer;
    }
    
    static void dfs(int index, int cnt, int sum, int target, int[] numbers){
        if(cnt == numbers.length){
            if(sum == target) answer++;
            return;
        }
        
        for(int i=index; i<numbers.length; i++){
            dfs(i+1, cnt+1, sum+numbers[i], target, numbers);
            dfs(i+1, cnt+1, sum-numbers[i], target, numbers);
        }
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
import kotlin.math.*
import java.util.*

var cnt = 0
var n = 0
val vis = Array(21){false}

class Solution {
    fun solution(numbers: IntArray, target: Int): Int {
        var answer = 0
        n = numbers.size
        dfs(numbers, 0, 0, target)
        
        return cnt
    }
    
    fun dfs(numbers:IntArray, index:Int, sum:Int, target:Int){
        if(n == index){
            if(sum == target){
                cnt++
            }
            return   
        }
        
        dfs(numbers, index+1, sum+numbers[index], target)
        dfs(numbers, index+1, sum-numbers[index], target)
    }
}
